### ISSUE TITLE
Random "Assertion `curr_LID!=-1' failed" error

### ISSUE THREAD

**Issue by Shangxin-Liu** [geodynamics/aspect#1831]
When we run some time-dependent simulations on our cluster, sometimes ASPECT randomly fails with this error:

_aspect: /home/shangxin/newriver/sources/trilinos-11.4.3-Source/packages/epetra/src/Epetra_BasicDirectory.cpp:633: int Epetra_BasicDirectory::GetDirectoryEntries(const Epetra_BlockMap&, int, const int_type*, int*, int*, int*, bool) const [with int_type = int]: Assertion `curr_LID!=-1' failed.
SIGABRT received
~_               

This behavior is very inconsistent - sometimes a simulation will run without error to the end time, then I'll start it again and it will fail with this error after a random time steps. The random nature actually makes me feel it's the issue of our system (sometime 'bad' nodes or others...), but before I ask our cluster administrator I'd like to get some ideas here.

The only possible consistent phenomenon I observe regarding this error is that for now the two cases having this error all fail at the tilmestep exactly after the "*** Snapshot created! " like this in the log file:

Snapshot created!

Timestep 400:  t=0.0195177 seconds
Primary job  terminated normally, but 1 process returned
a non-zero exit code.. Per user-direction, the job has been aborted.

I'm wondering if this problem has something to do with trillions? (I'm using trillions 11.4.3)

Btw, I saw on ASPECT mailing list last year there was already one message about this error on TACC stampede:

http://lists.geodynamics.org/pipermail/aspect-devel/2016-March/001241.html

But I didn't see more information regarding this question.

Any insights? @tjhei  @bangerth @gassmoeller

**Comment by bangerth** [comment]
Good question. I haven't seen anything like this before, but then I don't often run large simulations. I don't really know what to suggest -- we know that your cluster has had problems before, but I've also been long enough in the software business to know that it may nevertheless be a problem on our side. It is awkward for sure that the problems you see only happen occasionally. How many processors are involved in the computations where you see the error?

**Comment by Shangxin-Liu** [comment]
Hi Wolfgang,

Before I answer your question, I have a bunch of other observations regarding this issue.

Last night, I had several other cases failing with the similar issue but not exactly the same error message. I think we can firstly exclude the possibility of "out-of-memory" overflow because my test that has the total RAM 6-7 times larger than the DoFs limits also randomly runs into this error type. Although for now it's still a randomly happening issue, I have several consistent observations outlined here:

1. This error for now never happens on a single node parallel jobs. Each node on our clusters have 24 cores. For now I only randomly observed this error when I run jobs in more than one node (2,4,5, etc...)

2. Last night there are another two jobs failing with the similar issue but not the same error message. One case:

error file:
Error: the 1th element sent to 0 is greater than the total number of local elements: 212355 vs. 205920
n
Error: the 1th element sent to 0 is greater than the total number of local elements: 212355 vs. 196416
n
Error: the 1th element sent to 0 is greater than the total number of local elements: 212355 vs. 209088
n
Error: the 1th element sent to 0 is greater than the total number of local elements: 212355 vs. 205920
n
Error: the 1th element sent to 0 is greater than the total number of local elements: 212355 vs. 209088
n[nr029:71108] 4 more processes have sent help message help-mpi-api.txt / mpi-abort
[nr029:71108] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages

log file:

*** Snapshot created!

*** Timestep 1500:  t=0.0944297 seconds
   Solving temperature system... 7 iterations.
   Rebuilding Stokes preconditioner...-------------------------------------
MPI_ABORT was invoked on rank 2 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------

The other case:

error file: nothing

log file:
*** Snapshot created!

*** Timestep 3300:  t=0.191407 seconds
   Solving temperature system... 8 iterations.
   Rebuilding Stokes preconditioner...[nr033:22847] *** An error occurred in MPI_Wait
[nr033:22847] *** reported by process [140685869383681,1]
[nr033:22847] *** on communicator MPI COMMUNICATOR 3 DUP FROM 0
[nr033:22847] *** MPI_ERR_TRUNCATE: message truncated
[nr033:22847] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[nr033:22847] ***    and potentially your MPI job)

Although the error messages above are not the same as my first post, they all have a consistent pattern that the case gives out at the tilmestep exactly after the snapshot created. In these three cases I all set the parameter "Steps between checkpoint" to 100. However, after the failures last night, I reset this parameter to 500 to all the cases and resumed the simulation. I observed that for now none of them fails (Previously there must be some of my cases failing after several hours running at a tilmestep exactly after the snapshot created). So it seems that more frequently the snapshot is set to be created, the higher possibly the simulation gives out at a random tilmestep but consistently exactly after the snapshot created. This is the most consistent phenomenon I can provide regarding this random error for now:-)

I run all the simulation with a fixed mesh and I noticed that after each snapshot created, the simulation will rebuild Stokes preconditioner. I'm not sure how this related to snapshot and trillions.

3. The first "Assertion `curr_LID!=-1' failed" error message happened with 5 nodes, 120 cores test. The second and the third error types happened with 2 nodes, 48 cores test. All of the cases have the total DoFs limit from total cores RAM far larger than the simulation's DoFs so it seems not a "out of memory issue".

4. This random error does not influence the simulation results but I must always check my simulation every several hours and resume it if this random error happens....

5. Btw, as a side note, On our cluster commonly when the jobs fail with an error, it will automatically send an error message notification to my email. However, for this error, it never automatically reports to my email. It just crashes silently.... Our cluster administrator is very surprised by this behavior.

So any idea what's going on about this inconvenient random error@bangerth? And Timo @tjhei , did you get any further feedback from the mailing list regarding the similar error on TACC stampede last year? Also why this random error for now never happens on single node job further confuses me:-)

**Comment by Shangxin-Liu** [comment]
Some updates on this: We already confirmed that the only consistent pattern regarding this error is that when we lower the frequency to save snapshot (from every 100 to 500 timesteps), this error for now never happened on our cluster. However, when we changed back to every 100 timestep to save snapshots, some cases will start to fail randomly at the timestep exactly after the snapshot output. I noticed that snapshot saved during the process has a pretty large size. But the reason is still not clear to us and for now the only way we can do is to lower the snapshot saving frequency:-)   @bangerth @tjhei @gassmoeller
